### PR TITLE
[feature] filebackend support zerocopy & batch

### PR DESCRIPTION
**Pull request by Bruce-x-1997**
## Motivation

try to support zero_copy & batch interface in filebackend.

## Modifications

add `batch_set_v1` and `batch_get_v1` interface

## Benchmarking and Profiling
when I use my script to test interface performance.
```
test_performance_comparison_large_batch (__main__.TestHiCacheFilePerformanceComparison.test_performance_comparison_large_batch)
Compare performance with large batch (100 keys). ... 
================================================================================
Performance Comparison: Large Batch (100 keys)
================================================================================

[Old Interface] batch_set: 2.588 ms
[Old Interface] batch_get: 1.296 ms

[New Interface] batch_set_v1: 1.594 ms
[New Interface] batch_get_v1: 0.742 ms

================================================================================
Write Speedup: 1.62x (faster)
Read Speedup:  1.75x (faster)
================================================================================

ok
test_performance_comparison_medium_batch (__main__.TestHiCacheFilePerformanceComparison.test_performance_comparison_medium_batch)
Compare performance with medium batch (50 keys). ... 
================================================================================
Performance Comparison: Medium Batch (50 keys)
================================================================================

[Old Interface] batch_set: 1.237 ms
[Old Interface] batch_get: 0.636 ms

[New Interface] batch_set_v1: 0.807 ms
[New Interface] batch_get_v1: 0.362 ms

================================================================================
Write Speedup: 1.53x (faster)
Read Speedup:  1.76x (faster)
================================================================================

ok
test_performance_comparison_multiple_runs (__main__.TestHiCacheFilePerformanceComparison.test_performance_comparison_multiple_runs)
Compare performance with multiple runs for statistical accuracy. ... 
================================================================================
Performance Comparison: Multiple Runs (10 keys, 5 runs)
================================================================================

[Old Interface] Average batch_set: 0.249 ms
[Old Interface] Average batch_get: 0.136 ms

[New Interface] Average batch_set_v1: 0.177 ms
[New Interface] Average batch_get_v1: 0.078 ms

================================================================================
Average Write Speedup: 1.41x (faster)
Average Read Speedup:  1.73x (faster)
================================================================================

ok
test_performance_comparison_small_batch (__main__.TestHiCacheFilePerformanceComparison.test_performance_comparison_small_batch)
Compare performance with small batch (10 keys). ... 
================================================================================
Performance Comparison: Small Batch (10 keys)
================================================================================

[Old Interface] batch_set: 0.245 ms
[Old Interface] batch_get: 0.135 ms

[New Interface] batch_set_v1: 0.157 ms
[New Interface] batch_get_v1: 0.076 ms

================================================================================
Write Speedup: 1.56x (faster)
Read Speedup:  1.77x (faster)
================================================================================

ok
```
at the begining, there are some errors in my script, so I see 10x speedup
but now we can measure the most cost is in fileio, and at batch_set case, fileio has a higher weight
In order to utilize the parallel write capability, I still keeps "One file corresponds to one tensor"
as if I merge tensors in a file, all tensor must have to be written one by one, it will be slower
and if I use multi-thread, if the batch is not enough large, the overhead of asynchronous scheduling can become too large and slow things down.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/managers/cache_controller.py [hunk 1]
```diff
@@ -588,17 +588,38 @@ def append_host_mem_release(self, host_indices: torch.Tensor):
...
+
         inc = 0
+        failed_keys = []
         for i in range(len(hash_values)):
             if not results[i]:
-                logger.warning(
-                    f"Prefetch operation {operation.request_id} failed to retrieve page {hash_values[i]}."
-                )
+                failed_keys.append(hash_values[i])
+                if len(failed_keys) == 1:  # Only log first failure
```
Comment (pansicheng) [inline python/sglang/srt/managers/cache_controller.py]: It seems the original implementation can also achieve 'only log first failure' (it will exit the loop once it encounters `not results[i]`)

File: python/sglang/srt/mem_cache/hicache_storage.py [hunk 2]
```diff
@@ -272,3 +316,118 @@ def clear(self) -> bool:
...
+            )
+
+            # Stage 2: Get tensor from memory pool
+            page_start_idx = i * self.mem_pool_host.page_size
+            actual_idx = host_indices[page_start_idx].item()
+            page_tensor = self.mem_pool_host.get_data_page(actual_idx, flat=True)
+
+            # Stage 3: Make contiguous
+            if not page_tensor.is_contiguous():
+                page_tensor = page_tensor.contiguous()
```
Comment (pansicheng) [inline python/sglang/srt/mem_cache/hicache_storage.py]: `get_data_page` may produce non-contiguous tensors when [k, v] is the first dimension of `kv_buffer`. This can trigger unintended copies during `flatten()` or `contiguous()`, potentially resulting in incorrect copy destinations. To improve zero-copy efficiency, we could use address offsets for direct kv access

File: python/sglang/srt/managers/cache_controller.py (hunk 1 above)
Comment (Bruce-x-1997) [reply python/sglang/srt/managers/cache_controller.py]: done

File: python/sglang/srt/mem_cache/hicache_storage.py (hunk 2 above)
Comment (Bruce-x-1997) [reply python/sglang/srt/mem_cache/hicache_storage.py]: like what, memoryview?, or ctypes pointer direct changes to tensor?

Comment (Bruce-x-1997) [reply python/sglang/srt/mem_cache/hicache_storage.py]: done

File: python/sglang/srt/mem_cache/hicache_storage.py [hunk 3]
```diff
@@ -230,12 +318,30 @@ def batch_get(
...
+                # Stage 3: Convert to numpy view
+                buf = memoryview(target_location.view(torch.uint8).contiguous().numpy())
+
+                # Stage 4: File I/O
+                with open(tensor_path, "rb", buffering=0) as f:
+                    if f.readinto(buf) != expected:
+                        results.append(None)
+                        continue
+
+                results.append(target_location)
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/hicache_storage.py:369]: where did we write the loaded result from file into the target location?

Comment (Bruce-x-1997) [reply python/sglang/srt/mem_cache/hicache_storage.py:369]: what's your mean？

Comment (Bruce-x-1997) [reply python/sglang/srt/mem_cache/hicache_storage.py:369]: what's your mean

File: python/sglang/srt/managers/cache_controller.py [hunk 4]
```diff
@@ -588,9 +588,21 @@ def append_host_mem_release(self, host_indices: torch.Tensor):
...
     ):
+        """Zero-copy batch get: read data directly into host memory tensors."""
+        if not hash_values:
+            return
+
         results = self.storage_backend.batch_get_v1(
             hash_values, host_indices, extra_info
         )
+
+        if len(results) != len(hash_values):
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/cache_controller.py]: let's not have this kind of unnecessary branching on critical path, it's ok to put under debug level but not default

File: python/sglang/srt/managers/cache_controller.py [hunk 5]
```diff
@@ -599,6 +611,12 @@ def _page_get_zero_copy(
                 )
                 break
             inc += self.page_size
+
+        if inc == 0:
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/cache_controller.py]: same here, it's ok to have failed operation and the forward execution is designed to not be impacted

File: python/sglang/srt/managers/cache_controller.py [hunk 6]
```diff
@@ -702,6 +720,12 @@ def _storage_hit_query(self, operation) -> tuple[list[str], int]:
             if prefix_keys and len(prefix_keys) > 0:
                 prefix_keys += batch_hashes
 
+        if storage_query_count == 0:
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/cache_controller.py]: same here, no need to have it on critical path

File: python/sglang/srt/mem_cache/hicache_storage.py [hunk 7]
```diff
@@ -30,19 +31,6 @@ def get_hash_str(token_ids: List[int], prior_hash: str = None) -> str:
     return hasher.hexdigest()
 
 
-def hash_str_to_int64(hash_str: str) -> int:
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/hicache_storage.py:33]: please bring it back, as it breaks code some prior PR introduced

File: python/sglang/srt/mem_cache/hicache_storage.py
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/hicache_storage.py:184]: consider to move HiCacheFile into the `storage` directory as well since with zero-copy feature it becomes much more complicated

File: test/registered/hicache/hicache_utils/test_hicache_storage_performance_comparison.py [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+import time
+import unittest
+from typing import List, Tuple
+
+import torch
+
+from sglang.srt.mem_cache.hicache_storage import HiCacheFile, HiCacheStorageConfig
+
+
+class MockHostKVCache:
```
Comment (xiezhq-hermann) [inline test/registered/hicache/hicache_utils/test_hicache_storage_performance_comparison.py:29]: then we can have a benchmark file into the specific sub-directory as well, and no need to follow the test case structure, please have a readme describing using the benchmark file instead

Comment (Bruce-x-1997) [reply test/registered/hicache/hicache_utils/test_hicache_storage_performance_comparison.py:29]: ok, I have move to hicache_utils & add a readme, please review it ,thanks

File: python/sglang/srt/mem_cache/hicache_storage.py (hunk 7 above)
Comment (Bruce-x-1997) [reply python/sglang/srt/mem_cache/hicache_storage.py:33]: done

**Review by xiezhq-hermann — APPROVED** [review]

**Comment by xiezhq-hermann** [comment]
replacing it with a recent PR: https://github.com/sgl-project/sglang/pull/39880
